### PR TITLE
36978 | Fast image processor for DPT model

### PR DESCRIPTION
**Pull request by samrae7**
# What does this PR do?

Add Fast Image Processor for DPT model

Fixes #36978

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [Link to issue](https://github.com/huggingface/transformers/issues/36978)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+if is_torchvision_available():
+    if is_torchvision_v2_available():
+        from torchvision.transforms.v2 import functional as F
+    else:
+        from torchvision.transforms import functional as F
+
+
+def get_output_image_size_for_ensure_multiple(
+    # TODO type should be torch.Tensor
```
Comment (samrae7) [inline src/transformers/models/dpt/image_processing_dpt_fast.py:68]: ^

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+            # fit height
+            scale_width = scale_height
+
+    new_height = constrain_to_multiple_of(scale_height * input_height, multiple=multiple)
+    new_width = constrain_to_multiple_of(scale_width * input_width, multiple=multiple)
+
+    return (new_height, new_width)
+
+
+class DPTFastImageProcessorKwargs(DefaultFastImageProcessorKwargs):
```
Comment (samrae7) [inline src/transformers/models/dpt/image_processing_dpt_fast.py:110]: TODO: sort out documentation for these

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+        # Group images by size for further processing
+        # Needed in case do_resize is False, or resize returns images with different sizes
+        grouped_images, grouped_images_index = group_images_by_shape(resized_images)
+        processed_images_grouped = {}
+        for shape, stacked_images in grouped_images.items():
+            if do_center_crop:
+                stacked_images = self.center_crop(stacked_images, crop_size)
+            if do_pad:
+                stacked_images = torch.stack([self.pad_image(image, size_divisor) for image in stacked_images])
```
Comment (samrae7) [inline src/transformers/models/dpt/image_processing_dpt_fast.py:280]: Could do with some guidance here:

Is it much preferable to make pad_image handle batched tensors, and so avoid a list comprehension like this?

Is that for speed, ie. to take advantage of pytorch perfromance?

If so I can make that change quite quickly

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            do_pad=do_pad,
+            ensure_multiple_of=ensure_multiple_of,
+            keep_aspect_ratio=keep_aspect_ratio,
+            **kwargs,
+        )
+
+        data = {"pixel_values": processed_images}
+
+        if segmentation_maps is not None:
+            isList = False
```
Comment (samrae7) [inline src/transformers/models/dpt/image_processing_dpt_fast.py:195]: TODO: factor thios out into a method, _preprocess_segmentation_maps

File: src/transformers/models/dpt/image_processing_dpt_fast.py (hunk 3 above)
Comment (yonigozlan) [reply src/transformers/models/dpt/image_processing_dpt_fast.py:260]: Hey @samrae7 , yes that would be great!

**Comment by samrae7** [comment]
@yonigozlan I've looked at the test failures in the pipeline and they seem unrelated to my change. The branch was updated from upstream main just before the tests ran. Any advice on what to do?

**Comment by samrae7** [comment]
Update: I was on holiday for a week and can now pick this up again. There were apparently-unrelated tests failing in the pipeline. I will investigate

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+        elif not is_pil_image(segmentation_maps) and segmentation_maps.ndim == 2:
+            segmentation_maps = segmentation_maps.unsqueeze(0)
+
+        segmentation_maps = self._prepare_input_images(segmentation_maps)
+
+        if isBatched:
+            segmentation_maps = [map.squeeze(0) for map in segmentation_maps]
+        else:
+            segmentation_maps = segmentation_maps[0]
+        return segmentation_maps
```
Comment (yonigozlan) [inline src/transformers/models/dpt/image_processing_dpt_fast.py:304]: Could you have a look at how this is handled in the newly merged BeitImageProcessorFast?

**Review by yonigozlan — COMMENTED** [review]
Hey @samrae7 ! Thanks for iterating :). The Beit fast image processor was recently merged, and it seems quite similar to this one, could you have a look to try and uniformize this PR a bit based on it? Especially the handling of segmentation maps, as I'm not a fan of dragging them as ImageInput inside the `_preprocess` function

File: src/transformers/models/dpt/image_processing_dpt_fast.py (hunk 5 above)
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py:304]: Sure I'll do that, thanks 👍

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+            )
+        elif size.shortest_edge:
+            new_size = get_resize_output_image_size(
+                image,
+                size=size.shortest_edge,
+                default_to_square=False,
+                input_data_format=ChannelDimension.FIRST,
+            )
+        elif size.max_height and size.max_width:
+            new_size = get_image_size_for_max_height_width(image.size()[-2:], size.max_height, size.max_width)
```
[the remark this answers is not in the document]
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py]: removed this

File: src/transformers/models/dpt/image_processing_dpt_fast.py (hunk 5 above)
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py]: thanks for this suggestion. I've refactored based on the approach used in Beit Fast Processor

**Review by yonigozlan — COMMENTED** [review]
Thanks for iterating! I'd prefer having this be closer to Beit fast image processor, as the logic is almost the same

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        )
+        data = {"pixel_values": processed_images}
+
+        if segmentation_maps is not None:
+            segmentation_maps = make_list_of_images(images=segmentation_maps, expected_ndims=2)
+            segmentation_maps = self._preprocess_segmentation_maps(
+                segmentation_maps=segmentation_maps,
+                return_tensors=return_tensors,
+                **kwargs,
+            )
```
[the remark this answers is not in the document]
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py:176]: ok will make this and the other changes asap. thanks

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+        be overidden by `keep_aspect_ratio` in `preprocess`.
+    segmentation_maps (`ImageInput`, *optional*):
+        Segmentation map to preprocess.
+    """
+
+    ensure_multiple_of: Optional[int]
+    size_divisor: Optional[int]
+    do_pad: Optional[bool]
+    keep_aspect_ratio: Optional[bool]
+    segmentation_maps: Optional[ImageInput] = (None,)
```
[the remark this answers is not in the document]
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py]: added now thanks for spotting that!

File: src/transformers/models/dpt/image_processing_dpt_fast.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py]: done

File: src/transformers/models/dpt/image_processing_dpt_fast.py [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+
+@auto_docstring
+class DPTImageProcessorFast(BaseImageProcessorFast, SemanticSegmentationMixin):
```
[the remark this answers is not in the document]
Comment (samrae7) [reply src/transformers/models/dpt/image_processing_dpt_fast.py]: done. thanks

**Comment by samrae7** [comment]
@yonigozlan this is ready for re-review

**Review by yonigozlan — APPROVED** [review]
Hey @samrae7 ! Thanks a lot for working on this, looks great!
I made some final changes, mainly to use modular as a large part of the processing code was the same as Beit. Waiting for the CI then LGTM!
